Deprecate "array-as-vector" methods and adjust failing tests

dlongnecke-cray opened the pull request:
The following "array-as-vec" methods on arrays now have deprecation warnings inserted in them:

- push_back
- pop_back
- push_front
- pop_front
- clear
- insert
- remove

Deprecation warnings are off by default, but can be opted into by setting the config param `showArrayAsVectorWarnings` to `true`.

Tests that do not directly pertain to/are not about the deprecated methods have been modified to use the `list` type. When deprecation warnings are finally switched to be on, any tests written for "array-as-vector" methods will have to either:

- Be removed
- Moved to `test/deprecated`
- Have their `.good` files updated to contain the deprecation warnings

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none`
- [ ] `ALL` on `linux64` when `CHPL_COMM=gasnet`

e-kayrakli commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -1857,16 +1857,17 @@ module Sparse {
     const parentDT = transpose(D.parentDom);
     var Dom: sparse subdomain(parentDT) dmapped CS(sortedIndices=false);
-    Dom += indices;
+    Dom += indices.toArray();
```
More as a note going forward, but this makes performance of this method even worse then it was before. I am hoping to make this implementation more elegant and perform better with this: https://github.com/chapel-lang/chapel/pull/13420

dlongnecke-cray replied:
Do you think there's anything that can be done to soften the performance hit, short of just using the deprecated features?

vasslitvinov replied:
We can just iterate over `indices` and add each to `Dom` individually, right?

In any case, see what the now-replaced += did, and do the same here.

e-kayrakli replied:
Right, that's also an option.

`[i in indices] Dom += i;` should work, but it is hard to say whether that'd be faster or slower than this approach.

ben-albrecht replied:
I would advocate for @dlongnecke-cray not spending any additional effort modifying this implementation and accepting a temporary performance hit, with the assumption that @e-kayrakli will get around to writing the improved version with buffered index addition within the release cycle.

e-kayrakli replied:
@ben-albrecht -- totally agree

e-kayrakli commented on modules/standard/Lists.chpl:
```diff
@@ -54,13 +54,13 @@
 
   pragma "no doc"
-  config param _sanityChecks = true;
+  private config param _sanityChecks = true;
```
I know this is not related to this PR but since this is a `config`, should the name be more verbose? Even if it is not a user-facing feature..

dlongnecke-cray replied:
Thanks for the reminder. Actually, it makes no sense for it to be both private and config, since the concept of a "private config var" is inherently useless...

I'll probably just remove the `config` keyword.

e-kayrakli commented on test/library/packages/UnitTest/UnitTest.chpl:
```diff
@@ -1052,16 +1053,15 @@ module UnitTest {
           testStatus[testName] = true;
           return;
-        }
+        } catch {}
```
I think you have to catch a particular error from `indexOf` here.

mppf replied:
yeah, please never write `catch {}`. Even if that's what you think is right, put a comment explaining why the error can be safely ignored.

dlongnecke-cray replied:
This is really odd, but when I rewrite the try/catch to something like...

```
try { ... } catch e: IllegalArgumentError {} catch e { throw e; }
```

The `UnitTest` module tests fail with...

```
./UnitTest.chpl:1064: internal error: did not find the insertion point [resolution/lowerIterators.cpp:1883]
```

So for now, I'm just going to revert the `UnitTest` package and add an issue to rewrite it after the changes to the List API go through.

e-kayrakli commented on test/library/packages/UnitTest/UnitTest.chpl:
```diff
@@ -1073,10 +1073,9 @@ module UnitTest {
-            }
+            try {
+              checkCircle.remove(superTest:string);
+            } catch {}
```
same here

bradcray commented on modules/standard/Path.chpl:
```diff
@@ -182,25 +183,27 @@ proc commonPath(in paths: string ...?n): string {
-  var pos = prefixArray.size;   // rightmost index of common prefix
-  var minPathLength = prefixArray.size;
+  var pos = prefixList.size;   // rightmost index of common prefix
+  var minPathLength = prefixList.size;
```
What's the motivation for rewriting these idioms that use split() to create an array?  I'm not seeing calls to the newly-deprecated routines, but could easily be missing something.

dlongnecke-cray replied:
I believe that is because later on the function may make calls to `array.remove`, IE:

```
211   if (flag == 1) {
212     prefixArray.remove(pos..prefixArray.size);
213   } else {
214     prefixArray.remove(minPathLength+1..prefixArray.size);
215     // in case all paths are subsets of the longest path thus pos was never
216     // updated
217   }
```

lydia-duncan replied:
I wonder, should we consider having a version of `split` that returns a list (maybe by taking a param argument to flip between the return types or something)?  Or making `split` always return a list if we think that is the more common use case?

vasslitvinov replied:
Hm, then that wouldn't be specific to `split`, so I thought. However, this `split` here is an iterator, so perhaps the ideal way to code up this is to pass the iterator straight to a list initializer:
```chpl
var prefixList = new list(firstPath.split(pathSep, -1, false));
```

For that, the initializer would take an _iteratorRecord and iterate over it, like proc chpl__initCopy(ir: _iteratorRecord) at ChapelArray.chpl:4357:
```chpl
proc init(iterable: _iteratorRecord) {
  elemType = iteratorToArrayElementType(iterable.type);
  this.complete();
  for elm in iterable do
    this.append(elm);
}
```

vasslitvinov replied:
Separately, switching from array to `list` is not a good enough reason to rename a variable. In most cases, a variable's name has no business reflecting its type. So, `prefixArray` would be a perfectly fine name even if its type is `list`.

lydia-duncan replied:
> Separately, switching from array to list is not a good enough reason to rename a variable.

I disagree.  Maybe the type shouldn't be in the variable in the first place, but if there is a type in the name it *definitely* shouldn't be the wrong type.

bradcray replied:
I agree with Lydia and was logging in to say similarly.  If `myInt` ceases to have integer nature, we shouldn't leave the name the same to minimize diffs.  The current state of the code is ultimately more important than minimizing the diffs for a given PR.

vasslitvinov replied:
OTOH we do not have to interpret the name `prefixArray` to mean literally "its type is a Chapel array". We can interpret this name as "this is a collection that looks like an array". For which `list` is a perfectly legitimate implementation.

Regardless, I am OK with changing the name to `prefixList`.

bradcray commented:
One way to get good review on large PRs like this is to have people who have traditionally "owned" a piece of code review those pieces.  This has a few major benefits:

* It spreads out the pain of reviewing an otherwise large PR
* It gets the people who know the code best reviewing the changes.  This should make it go faster.  It also heads off surprises down the road when someone notices and says "you changed what to do what?"
* It gets more people seeing the practical consequences of what a change like this will be (in code that they're presumably invested in to some degree), which will likely result in good feedback on the change (as compared to the silence that often results from "who wants to review my design" mails

For example, as I was looking through the files in question, I might suggest the following ownership mapping:

* @ben-albrecht review the LinearAlgebra.chpl, vectorMethods.chpl, testSparseCS.chpl, and testDate.chpl changes
* @lydia-duncan review the Path.chpl changes
* @mppf review the Spawn.chpl, test/classes/delete-free, test/library/packages/Sort/*, and test/studies/dedup/*, test/*/ferguson/ changes
* @benharsh review the test/*/bharshbarg/ changes
* @daviditen review the LCALS, arrayVectorOps.chpl, and test/*/diten/ changes
* @bradcray review the test/arrays/errors/, test/exercises/c-ray, and test/arrays/userAPI/ changes
* @vasslitvinov review the test/library/packages/VectorList/* changes

I mentioned to you on chat, but I also worry a bit about the number of cases that we're kicking down the road by adding deprecation warnings to their .good files (since we'll have to come back to these eventually when we actually remove the routines).  As people were reviewing their specific pieces, they could make the call as to whether the file should be rewritten or moved to the test/deprecated/ directory (if it no longer has a role in the new world and really should just be removed once those routines go away).

I also mentioned this on chat, but I wonder if we should postpone adding the deprecation warnings themselves until such time as TOML and mason are ready for them since they account for a large number of the file changes (and will need to be undone once those pieces are rewritten).

dlongnecke-cray commented:
Thanks for holding my hand and walking me through it, this time. Will definitely help the next time I ever have to do a large/sweeping PR. 

I see imminent TODO as being the following:

- [ ] Remove deprecation warnings
- [ ] Remove `.compopts` with `--no-warnings` only behavior.
- [ ] Do another pass over tests, marking (at least internally) tests that should be moved to `deprecated` or otherwise be removed.

Generally, I have tried to avoid updating the `.good` file for a test unless its uses of "soon-to-be-deprecated" methods are self contained, or it makes no sense without the deprecated method calls. The former may be candidates for rewriting, the latter are best suited for moving to `test/deprecated`. 

I will do a quick sweep to try and categorize `.good` file changes after I do the above.

mppf commented on test/classes/delete-free/lifetimes/bad-push-back-borrow.good:
```diff
@@ -1,4 +1,5 @@
 bad-push-back-borrow.chpl:20: In function 'addone':
+bad-push-back-borrow.chpl:25: warning: push_back is deprecated - please use List.extend
```
These lifetime push-back tests should be updated to test `list`

dlongnecke-cray replied:
Hmmm, I duplicated this test but using `list` instead, and gave `list.append` the following new signature:

`proc append(pragma "no auto destroy" in: x) lifetime this < x { ... }`

But the test compiles just fine instead of producing an error.

mppf replied:
OK, then just make the `list` version a `.future` and create a TODO issue to revisit it and assign it to me.

mppf commented on test/library/packages/TOML/test/TomlTest.compopts:
```diff
@@ -0,0 +1,9 @@
+# Instead, we suppress all compiler warnings (until TOML is rewritten to not
+# use "array-as-vec" operations).
+#
+# See: https://github.com/chapel-lang/chapel/issues/13233
```
I would expect that TOML is just using 1-D arrays and so could easily switch to using `list` ?

lydia-duncan commented on modules/standard/Path.chpl:
```diff
@@ -535,6 +549,7 @@ private proc normalizeLeadingSlashCount(name: string): int {
   :rtype: `string`
 */
 proc normPath(name: string): string {
+  use Lists;
```
Since you add this use to more than three functions, I think it would be good to instead add a module-level `private use`

lydia-duncan commented on modules/standard/Path.chpl:
```diff
@@ -554,13 +569,13 @@ proc normPath(name: string): string {
   }
 
-  var result = pathSep * leadingSlashes + pathSep.join(outComps);
+  var result = pathSep * leadingSlashes + pathSep.join(outComps.toArray());
```
It seems worth adding an overload of `join` for lists

mppf replied:
Or for something "iterable" which maybe one day will have a reasonable definition :)

bradcray replied:
> Or for something "iterable"

(and could be implemented today via _iteratorRecord even though it's not user-facing... is that what you're implying?)

mppf replied:
Yeah, I think that today _iteratorRecord + array might cover most cases, but even there I am not so sure (because ultimately you'd want to be able to pass in a record/class that had a `proc these` but I don't think we turn that into an _iteratorRecord today....). So in practice it probably means leaving the argument totally generic and looping over it with `for`.

mppf replied:
> maybe one day will have a reasonable definition

this something we can expect to get with concepts/interfaces

vasslitvinov replied:
Aha, others thought about it before me. I conjecture that `sep.join(outComps.these())` would do it.

vasslitvinov commented on modules/internal/ChapelArray.chpl:
```diff
@@ -167,6 +167,9 @@ module ChapelArray {
   use ArrayViewReindex;
 
+  pragma "no doc"
+  config param showArrayAsVectorWarnings = false;
```
Is this a missed change? The remainder of this file uses `showArrayAsVectorDeprecations`. Maybe that's why the checks are failing at the moment.

vasslitvinov replied:
Also, is it supposed to be `false` by default?

bradcray replied:
The original PR had the checks always-on.  The config was added so programmers could opt-in / out.  The goal is to release with this `true` but this PR may be merged with it in the `false` condition if we don't have clean testing (if we do, it should be changed back to `true`).

dlongnecke-cray replied:
One thing that is still yet to be decided is what to do with all the tests written specifically for "array-as-vec" operations. If we want to merge this PR with the config ON, then we will have to do one of the following for all those tests (where _currently_ they are unchanged):

- Update their good files
- Remove the test
- Migrate the test to `test/deprecated`

mppf replied:
My 2¢ for tests specifically for array-as-vec: update their good files now, and when the deprecated code is removed, make sure that tests of functionality that exists in `list` are moved over to `list`.

daviditen replied:
I agree with @mppf. The ideal thing to do in the end is to update the tests to do what they do now, but using `list` functionality.

bradcray replied:
I disagree that we should put off rewriting tests to use list until the array-is-vec features are removed.  If the tests exercise array-is-vec functionality that also exists on list, I think those tests should be rewritten now to put more weight on list since that's the preferred solution and the one we want to make sure is at least as good as array-is-vec was.  If existing list tests already covers the test's functionality sufficiently that the rewriting is unnecessary, I think the test should just be moved to test/deprecated (with updated .good file).  If the test simply doesn't make sense in a list world (e.g., queries the domain of an array after a number of array-as-vec operations), I'd also move it to test/deprecated.

vasslitvinov commented on modules/standard/Path.chpl:
```diff
@@ -209,13 +212,16 @@ proc commonPath(in paths: string ...?n): string {
     // updated
   }
 
+  var prefixArray = prefixList.toArray();
```
No need to convert to an array. There is an overload of string.join that takes an iterable. So I expect the following to work:
```chpl
return pathSep.join(prefixList.these()); 
```

Ideally, `these()` would not be required, however we are not there at the moment.

vasslitvinov commented on modules/standard/Path.chpl:
```diff
@@ -260,25 +267,29 @@ proc commonPath(paths: []): string {
-  var prefixArray = firstPath.split(delimiter, -1, false);
+  var prefixList = new list(string);
+  for x in firstPath.split(delimiter, -1, false) do
+    prefixList.append(x);
```
Same comments about variable naming and initialization. Variable naming is much of a matter of taste, so up to you.

vasslitvinov commented on modules/standard/Path.chpl:
```diff
@@ -287,13 +298,16 @@ proc commonPath(paths: []): string {
   }
 
+  var prefixArray = prefixList.toArray();
   result = delimiter.join(prefixArray);
```
The same comment about .toArray() vs. .these().

vasslitvinov commented on modules/standard/Path.chpl:
```diff
@@ -690,16 +707,19 @@ proc relPath(name: string, start:string=curDir): string throws {
     return curDir;
 
-  return joinPath(outComps);
+  return joinPath(outComps.toArray());
```
Same comments about initializing `outComps` and about `.toArray`.

vasslitvinov commented on modules/standard/Spawn.chpl:
```diff
@@ -655,13 +655,18 @@ module Spawn {
-    if shellarg != "" then args.push_front(shellarg);
-    args.push_front(executable);
-    return spawn(args, env, executable,
+    return spawn(args.toArray(), env, executable,
```
Another way to write this is:
```chpl
var args = if shellarg == "" then [executable, command] else [executable, shellarg, command];
```

vasslitvinov commented on modules/internal/ChapelArray.chpl:
```diff
@@ -3002,6 +3005,8 @@ module ChapelArray {
      */
     proc push_back(in val: this.eltType) lifetime this < val {
+      if showArrayAsVectorDeprecations then
+        compilerWarning("push_back is deprecated - please use List.extend");
```
I expect the replacement to be `append`, not `extend`. Could you double-check?

vasslitvinov commented on test/library/packages/VectorList/test-ints.good:
```diff
@@ -1,3 +1,5 @@
+test-ints.chpl:20: In function 'help':
+test-ints.chpl:25: warning: push_back is deprecated - please use List.extend
```
I can fix this  -- once this PR is merged.

benharsh commented on test/arrays/bharshbarg/popFrontLeak3.good:
```diff
@@ -1,3 +1,6 @@
+popFrontLeak3.chpl:37: In function 'main':
+popFrontLeak3.chpl:42: warning: push_back is deprecated - please use List.extend
+popFrontLeak3.chpl:43: warning: pop_front is deprecated - please use List.pop
```
I think it would be worth rewriting these popFrontLeak[1-3] tests to use the List type to ensure that the leak wasn't accidentally re-introduced.

dlongnecke-cray replied:
I've made an issue to remind myself to do this after `list` gets support for `init=` from an `_iteratorRecord`.

benharsh commented on test/optimizations/bulkcomm/bharshbarg/stridedAssignDR.good:
```diff
@@ -0,0 +1,45 @@
+stridedAssignDR.chpl:25: In function 'buildSlices':
```
This test should be updated to use the new type. I can help you figure out what this program is doing, or I can make the change myself post-merge.

dlongnecke-cray replied:
I've made an issue for rewriting this here: https://github.com/cray/chapel-private/issues/374

benharsh commented on test/studies/shootout/reverse-complement/bharshbarg/revcomp-line.good:
```diff
@@ -1,3 +1,6 @@
+revcomp-line.chpl:71: In function 'readUntil':
+revcomp-line.chpl:80: warning: push_back is deprecated - please use List.extend
+revcomp-line.chpl:83: warning: push_back is deprecated - please use List.extend
```
These two reverse-complement should probably be rewritten to use something like the ``bytes`` type instead of ``List``, so these deprecation warnings are fine for now.

ben-albrecht commented on test/sparse/CS/multiplication/correctness.chpl:
```diff
@@ -126,7 +128,7 @@ proc multiply(A: [?ADom] ?eltType, B: [?BDom] eltType) where isSparseArr(A) && i
   }
 
-  var indices: [1..0] 2*int;
+  var indices: list(2*int);
```
@e-kayrakli - I believe this test could also benefit from buffered index addition, however it's not incredibly important for us to keep performance optimal for this test.

ben-albrecht commented on test/release/examples/primers/arrayVectorOps.good:
```diff
@@ -1,3 +1,17 @@
+arrayVectorOps.chpl:29: warning: push_front is deprecated - please use List.insert
```
I think we should consider removing these examples from the primer, or at least adding a `.. warning ::` / `.. note::` that vector operations are being deprecated on arrays and users should refer to the Lists module for support of vector operations.

dlongnecke-cray replied:
I've made an issue to keep track of this. The primers will be updated when deprecation warnings are switched on by default.

bradcray commented on test/arrays/errors/domainTypeAsDomainExpr.good:
```diff
@@ -1,2 +1,3 @@
+domainTypeAsDomainExpr.chpl:11: warning: push_back is deprecated - please use List.extend
```
@dlongnecke-cray: For this test, the use of push_back is completely independent of the purpose of the test (I think it was inherited from some code a user sent in).  Would you mind editing domainTypeAsDomainExpr.chpl  as follows:

```diff
-var x: [0..-1] owned Child;
-x.push_back(new owned Child("C"));
+var x: [0..0] owned Child = new owned Child("C");
```

(otherwise, I can do this in a separate PR post-merge).

(Pro tip: Saying the above reminds me of another strategy for tackling big changes to the test system like this:  check in an ability for developers to opt into the new behavior — like the config param you've added — check that in, list the tests that need to be dealt with, find owners for each (as I did here w.r.t. the reviews), and then have people rewrite the tests themselves to spread out the effort; then once everyone's done their bit, flip the config param bit on master.  This tends to work best when deadlines are further away and people are feeling most calm).

dlongnecke-cray replied:
Sounds good, I will admit I was a bit confused about that code at the time.

bradcray commented:
I've looked at the tests I assigned to myself above.  c-ray looks good.  I commented on one test that I think should be updated.  I also think that the test/arrays/userAPI/ tests should be significantly simplified, but it's sufficiently complicated to explain that I can just take that on in a follow-on PR once this is merged.

dlongnecke-cray commented:
> > @ben-albrecht review the LinearAlgebra.chpl, vectorMethods.chpl, testSparseCS.chpl, and testDate.chpl changes
> 
> I took a pass through these tests and made a few additional comments.
> 
> My only requested change is that we mention array vector ops are being deprecated in the [array vector ops primer](https://chapel-lang.org/docs/primers/arrayVectorOps.html) (though I could be convinced that we should do that in a separate PR).
> 
> At a high-level, we generally place tests containing deprecated features into the `tests/deprecated` as staging to be removed in future releases. However, I can see how this deprecation can be considered a special case, and wouldn't be worth the extra effort of doing so.
> 
> Also, when we eventually remove these features from the language, do we intend to remove all the tests that currently contain deprecation warnings?
> 
> Thanks!

@ben-albrecht 

I think the plan here is to start with deprecation warnings being "opt in" via a config var. Closer to release, we will turn them "on", at which point all the tests _on_ "array-as-vec" operations will need to be either:

- Removed
- Moved to `test/deprecated` (a few choice tests, perhaps)
- Have their `.good` files updated (and probably moved)

When the time comes to fully deprecate "array-as-vec" operations, cornerstone tests will be added to `test/deprecated` for that purpose.

dlongnecke-cray commented:
Nope, that's a bad rebase...not going to bother!

dlongnecke-cray commented:
I figured I would clean up the commit history since I meandered around a fair bit after talking with Brad, but I'm not sure about the best way to do that. Clearly, I rebased incorrectly with that second to last force push...

vasslitvinov commented:
@dlongnecke-cray the "Merge pull request" button has the option "Squash and merge". This is what many of us have used in the past. Be sure that the PR number i.e. #13530 appears in the first line of the commit message. See ex. these merges: 04a74bc5c9 c5aeaa58ea 15eab334ca etc.

LouisJenkinsCS commented:
May I interject and suggest that the deprecation should not be pushed until #13652 is resolved in such a way that `list` outperforms `push_back` with `--local --fast`?

bradcray commented:
Just to make sure there's no confusion (or that I'm not the one confused): This PR doesn't actually deprecate the routines, right?  (in that the deprecation warning is still off by default).

dlongnecke-cray commented:
This PR does not actually deprecate the routines, and if it does I will patch that immediately.

dlongnecke-cray commented:
I have updated the OP with the final PR message.

bradcray commented:
It was the PR's title that was making me confused.

dlongnecke-cray commented:
Ah. There's always a mistake. I wrote such a beautiful PR message, but then forgot to change the title.
